src/packet_analysis/protocol/ethernet/Ethernet.cc

```diff
@@ -56,31 +53,25 @@ bool EthernetAnalyzer::AnalyzePacket(size_t len, const uint8_t* data, Packet* pa
 	// Other ethernet frame types
 	if ( protocol <= 1500 )
 		{
-		if ( 16 >= len )
+		len -= 14;
+		data += 14;
+
+		if ( len < protocol )
 			{
 			Weird("truncated_ethernet_frame", packet);
 			return false;
 			}
 
 		// Let specialized analyzers take over for non Ethernet II frames.
-		// Note that pdata remains at the start of the ethernet frame.
-
-		AnalyzerPtr eth_analyzer = nullptr;
-
-		if ( data[14] == 0xAA && data[15] == 0xAA )
+		if ( data[0] == 0xAA && data[1] == 0xAA )
```

Comment:
Couldn't this have leveraged the existing `DetectProtocol()` infrastructure reasonably well without introducing artificial  protocol identifiers? That would've been a bit more concise? The code here and the repeat in VLAN.cc is `DetectProtocol()` for SNAP/Novell in disguise, no? :-)

Not sure about the LLC default analyzer part, but that maybe just works (?) or may need some massaging to default to it within the packet-analysis infrastructure.

Reply:
🙈 indeed `DetectProtocol()` looks like it was made for exactly that use case...

Reply:
@awelzel and I talked about this a bit. I like the idea of using `DetectProtocol()` except the function has to return true or false if it matches some heuristic. There isn't a heuristic for LLC. It could always just return true, but that means the check in `ForwardPacket` is always going to hit that case and try to call `AnalyzePacket` on the LLC analyzer.

A workaround for calling it from the ethernet analyzer would be to make `DetectProtocol` check `packet->eth_type`. Unfortunately, that value isn't set from the VLAN analyzer. In that case it'd be set to 0x8100 since that's what the ethernet analyzer set it to before the VLAN analyzer got called.

I could pull the protocol detection bits out of `ForwardPacket`, call that from the analyzers, and then call the LLC analyzer if it doesn't match, but then the Ethernet/VLAN analyzers have to have a copy of the LLC analyzer directly. That violates how we're trying to keep to the analyzers from really knowing about each other beyond the registrations (and basically adds back in one of the hacks that this PR originally removed).

Reply:
> ... and there isn't a heuristic for LLC

Wouldn't that just be the inverse of SNAP and Novell, i.e. payload neither starts with 0xFFFF nor 0xAAAA? The problem here is that the heuristics might trigger for unknown Ethernet II types as well. One could argue that we are dealing with two protocols in one and should split the analyzer again. In that case LLC could become the default analyzer for non-Ethernert II.

Reply:
> The problem here is that the heuristics might trigger for unknown Ethernet II types as well.

Yes, exactly.

> One could argue that we are dealing with two protocols in one and should split the analyzer again. In that case LLC could become the default analyzer for non-Ethernert II.

So what, you'd have an analyzer that gets called for Ethernet II and one that gets called for everything else? Would that also split the VLAN analyzer into two?

Reply:
> So what, you'd have an analyzer that gets called for Ethernet II and one that gets called for everything else?

Yes, that's what I had in mind. I guess we would have a generic Ethernet analyzer that just extracts src and dst MACs, one analyzer for Ethernet II and one Ethernet multiplexer.  The Ethernet II and Ethernet multiplexer analyzers would use `DetectProtocol` to attach based on the Ethertype/length value. The Ethernet multiplexer could then use the first two payload bytes for classic dispatching (0xFFFF, 0xAAAA, and default for LLC):

```
          [DetectProtocol]        [Identifier]

                 ┌► EthernetII         ┌─► Novell
GenericEthernet ─┤                     │
                 └► EthernetMultiplex ─┼─► SNAP
                                       │
                                       └─► LLC (default)
```

> Would that also split the VLAN analyzer into two?

Forwarding from the VLAN analyzer would be done the same way it's done in the generic Ethernet analyzer: Use `DetectProtocol` to distinguish between Ethernet II and the rest.

All in all, these would be a lot of indirections. In particular, this would add an additional dispatching step to the probably most common Ethernet II path. My feeling is still that introducing child packet analyzers would be a bit cleaner. When called, a child analyzer would consume a number of bytes and just return how many bytes it processed to the calling analyzer. The calling analyzer could then continue processing at this point. VLAN would be the child in this case. I think I vaguely remember a draft for packet analyzers in which `ForwardPacket` returned a pair indicating parsing success and bytes left (that might be a textbook example of a monad ;). This way the "child-analyzer-concept" would just blend in.